_posts/news/2021-02-12-announcing-scalajs-1.5.0.md

```diff
@@ -0,0 +1,143 @@
+
+{% highlight scala %}
+trait Outer extends js.Object {
+  object InnerObject extends js.Object
+
+  class InnerClass extends js.Object
+}
+{% endhighlight %}
+
+While the above definition would be allowed by the compiler, any attempt to actually extend `Outer` would result in broken code.
+
+Scala.js 1.5.0 will therefore reject this definition at compile-time.
+This may cause code that compiled before to be rejected, which constitutes a backward source incompatible change.
+
+If you have actual, working code using that pattern (for example, where you receive instances of `Outer` from JavaScript code, instead of extending it yourself), it is likely that you can fix the code by adding `@js.native` both to the outer trait and the inner classes and objects:
+
+{% highlight scala %}
+@js.native
+trait Outer extends js.Object {
+  @js.native
+  object InnerObject extends js.Object
+
+  @js.native
+  class InnerClass extends js.Object
+}
+{% endhighlight %}
+
+## Custom JS function types
+
+**Scala 2.11 note:** With Scala 2.11, this feature requires the `-Xexperimental` flag, which adds SAM support to the Scala compiler.
+
+For interoperability with JavaScript code that manipulates function values, Scala.js has always had a set of `js.FunctionN` and `js.ThisFunctionN` types.
+Since they are SAM types, they can be created with anonymous functions, for example:
+
+{% highlight scala %}
+val jsFun: js.Function1[Int, String] = arg => arg.toString()
+{% endhighlight %}
+
+The built-in JS function types can accomodate any function with a fixed number of parameters, up to 22.
+However, they are not enough for more specialized use cases:
+
+* Function values with variadic arguments (varargs)
+* Function values of more than 22 arguments
+
+To address such advanced use cases, Scala.js 1.5.0 now allows to define custom JS function types with arbitrary `apply` signatures.
+For example, if you need to represent a function value with one `Int` parameter and varargs of type `String`, you can define a custom JS function type for it:
+
+{% highlight scala %}
+trait SpecialJSFunction extends js.Function {
+  def apply(first: Int, rest: String*): Any
+}
+{% endhighlight %}
+
+and create values of that type as expected:
+
+{% highlight scala %}
+val specialJSFun: SpecialJSFunction = { (first, rest) =>
+  s"$first, ${rest.size}"
+}
+{% endhighlight %}
+
+A custom JS function type is any `trait` whose parent class is `js.Function`, and which has a single abstract method named `apply`.
+If a custom JS function type extends the special trait `js.ThisFunction`, its first argument on the Scala.js side maps to the `this` value in JavaScript.
+
+Note that the built-in `js.FunctionN` and `js.ThisFunctionN` types fit those definitions, and are therefore "custom" JS function types themselves, that happen to be defined in the Scala.js standard library.
+There are no more built-in function types.
```

Comment:
I do not understand what this sentence means. Are you trying to say that there are not custom function types in the Scala.js language per-se? If yes, I think the previous sentence already does this sufficiently well.

Reply:
Yes, that was it. I removed the sentence.